File: RegexNfa/Infrastructure/Transition.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace RegexNfa.Infrastructure
{
    public class Transition
    {
        public string Id { get; }
        public State FromState { get; set; }
        public State ToState { get; set; }
        public char Atom { get; set; }

        public Transition()
        {
            Id = Guid.NewGuid().ToString();
            FromState = new State();
            ToState = new State();
        }

        public Transition(State fromState, State toState, char atom = (char)0) :
            this()
        {
            FromState = fromState;
            ToState = toState;
            Atom = atom;
        }

        public JObject ToJson()
        {
            JObject json = new JObject(
                new JProperty("Id", Id),
                new JProperty("Atom", Atom.ToString()),
                new JProperty("FromState", FromState.Id),
                new JProperty("ToState", ToState.Id));
            return json;
        }

        public override string ToString()
        {
            return this.ToJson().ToString();
        }
    }
}

File: src/transition.rs
use super::state::State;

#[derive(Debug)]
pub struct Transition<'a> {
    atom: char,
    from_state: &'a State<'a>,
    to_state: &'a State<'a>,
}

impl<'a> Transition<'a> {
    pub fn new(atom: char, from_state: &'a State, to_state: &'a State) -> Transition<'a> {
        Transition {
            atom,
            from_state,
            to_state,
        }
    }
}
